### PR TITLE
feat: implement "stateless" modules

### PR DESCRIPTION
**Pull request by charles-cooper**
resolves https://github.com/vyperlang/vyper/issues/2431, https://github.com/vyperlang/vyper/issues/2670, https://github.com/vyperlang/vyper/issues/3294

partially implements `.vyi` files https://github.com/vyperlang/vyper/issues/2814 (but does not handle ellipsis for bytestring/dynarray length, that will come in a follow up PR)

partially finishes https://github.com/vyperlang/vyper/pull/2852 (although it does not automatically export events in the ABI)

### Commit message

```

this commit implements support for "stateless" modules in vyper.

this is the first major step in implementing vyper's module system.
it redesigns the language's import system, allows calling internal
functions from imported modules, allows for limited use of types from
imported modules, and introduces support for `.vyi` interface files.

note that the following features are left for future, follow-up work:
- modules with variables (constants, immutables or storage variables)
- full support for imported events (in that they do not get exported
  in the ABI)
- a system for exporting imported functions in the external interface
  of a contract

this commit first and foremost changes how imports are handled in vyper.

previously, an imported file was assumed to be an interface file. some
very limited validation was performed in `InterfaceT.from_ast`, but not
fully typechecked, and no code was generated for it.

now, when a file is imported, it is checked whether it is
  1. a `.vy` file
  2. a `.vyi` file
  3. a `.json` file

the `.json` pathway remains more or less unchanged. the `.vyi` pathway
is new, but it is fairly straightforward and is basically a "simple"
path through the `.vy` pathway which piggy-backs off the `.vy` analysis
to produce an `InterfaceT` object.

the `.vy` pathway now does full typechecking and analysis of the
imported module. some changes were made to support this:
- a new ImportGraph data structure tracks the position in the import
  graph and detects (and bands) import cycles
- InputBundles now implement a `_normalize_path()` method.
  this method normalizes the path so that source IDs are stable no
  matter how a file is accessed in the filesystem (i.e., no matter
  what the search path was at the time `load_file()` was called).
- CompilerInput now has a distinction between `resolved_path` and `path`
  (the original path that was asked for). this allows us to maintain UX
  considerations (showing unresolved paths etc) while still having a
  1:1:1 correspondence between source id, filepath and filesystem.

these changes were needed in order to stabilize notions like "which file
are we looking at?" no matter the way the file was accessed or how it
was imported. this is important so that types imported transitively
can resolve as expected no matter how they are imported - for instance,
`x.SomeType` and `a.x.SomeType` resolving to the same type.

the other changes needed to support code generation and analysis for
imported functions were fairly simple, and mostly involved generalizing
the analysis/code generation to type-based dispatch instead of AST-based
dispatch.

other changes to the language and compiler API include:
- import restrictions are more relaxed - `import x` is allowed now
  (previously, `import x as x` was required)
- change function labels in IR
  function labels are changed to disambiguate functions of the same name
  (but whose parent module are different). this was done by computing a
  unique function_id for every function and using that function_id when
  constructing its IR identifier.
- add compile_from_file_input which accepts a FileInput instead of a
  string. this is now the new preferred entry point into the compiler.
  its usage simplifies several internal APIs which expect to have
  `source_id` and `path` in addition to the raw source code.
- change compile_code api from contract_name= to contract_path=

additional changes to internal APIs and passes include:
- remove `remove_unused_statements()`
  the "unused statements" are now important to keep around for imports!
  in general, it is planned to remove both the AST expansion and
  constant folding passes as copying around the AST results in both
  performance and correctness problems
- abstract out a common exception rewriting pattern.
  instead of raising `exception.with_annotation(node)` -- just catch-all
  in the parent implementation and then don't have to worry about it at
  the exception site.
- rename "type" metadata key on most top-level declarators to more
  specific names (e.g. "func_type", "getter_type", etc).
- remove dead package pytest-rerunfailures
  use of `--reruns` was removed in c913b2db0881a6
- refactor: move `parse_*` functions, remove vyper.ast.annotation
  move `parse_*` functions into new module vyper.ast.parse and merge in
  vyper.ast.annotation
- rename the old `GlobalContext` class to `ModuleT`
- refactor: move InterfaceT into `vyper/semantics/types/module.py`
  it makes more sense here since it is closely coupled with `ModuleT`.
```
### Cute Animal Picture

image

### CODE REVIEW

**Review by fubuloubu — COMMENTED** [review]
Address security notes and more manual testing from someone else

**Comment by charles-cooper** [comment]
> Address security notes and more manual testing from someone else

they were mostly just notices about previously existing import cycles. i've dismissed them

**Comment by charles-cooper** [comment]
re. manual testing --do you have someone specific in mind? not sure we will get much manual testing anyway until it's merged and people have a build / specific commit to go off of

File: vyper/semantics/types/function.py
Comment (DanielSchiavini) [inline vyper/semantics/types/function.py:51]: Sounds like https://github.com/vyperlang/titanoboa/pull/108 :smile:

File: vyper/semantics/types/module.py
Comment (DanielSchiavini) [inline vyper/semantics/types/module.py:69]: What is this supposed to mean?

File: tests/unit/compiler/asm/test_asm_optimizer.py
Comment (charles-cooper) [reply tests/unit/compiler/asm/test_asm_optimizer.py:74]: not yet - i was going to add!

File: tox.ini
Comment (charles-cooper) [reply tox.ini:56]: that's what i have in titanoboa - it makes type signatures less verbose, and i haven't found mypy warning about kwargs being optional (which is where it's usually used) to have been super helpful

File: vyper/__init__.py
Comment (charles-cooper) [reply vyper/__init__.py:3]: i think we might actually be able to get rid of the F401 thing somehow - maybe if we export explicitly via `__all__`

File: vyper/ast/parse.py
Comment (charles-cooper) [reply vyper/ast/parse.py:38]: it's used for constructing special builtins - https://github.com/vyperlang/vyper/blob/10564dcc37756f3d3684b7a91fd8f4325a38c4d8/vyper/builtins/_utils.py#L17

(this machinery should go away soon now btw that we can implement a stdlib using native vyper instead of being compiler intrinsics)

File: vyper/builtins/interfaces/ERC20.vyi
Comment (charles-cooper) [reply vyper/builtins/interfaces/ERC20.vyi:3]: no, i don't think so -- sender is only used for calls from boa. also, events are constructed using positional arguments only anyways.

File: vyper/codegen/function_definitions/common.py
Comment (charles-cooper) [reply vyper/codegen/function_definitions/common.py:50]: `function_id` is an int - it can be `0`! also i think i just have a habit of using explicit bools instead of truthy values as branch conditions :P

File: vyper/codegen/stmt.py
Comment (charles-cooper) [reply vyper/codegen/stmt.py:125]: there is magic in the AST copier which avoids copying fields which start with `"_"`. i actually want to get rid of the metadata dict and replace it with real members (i started on this by adding `_expr_info` to all exprs https://github.com/vyperlang/vyper/blob/10564dcc37756f3d3684b7a91fd8f4325a38c4d8/vyper/ast/nodes.py#L747 but they haven't been populated yet), but it's a bit of a longer-term project

File: vyper/semantics/types/function.py
Comment (charles-cooper) [reply vyper/semantics/types/function.py:51]: yep it's similar :)

File: vyper/semantics/types/module.py
Comment (charles-cooper) [reply vyper/semantics/types/module.py:69]: i think at some point i wanted to reverse the order of the arguments. so the API looks like `other.implements(x)` as opposed to `x.implements(other)`

File: vyper/codegen/stmt.py
Comment (DanielSchiavini) [reply vyper/codegen/stmt.py:125]: Moving the calls to a single function might be useful then, like `get_metadata(node, 'key')` or `get_type(node)`

File: vyper/__init__.py
Comment (DanielSchiavini) [reply vyper/__init__.py]: sounds good and clear :+1:

File: tox.ini
Comment (DanielSchiavini) [reply tox.ini:56]: I misunderstood the flag, adding optional[x]=None is not really needed indeed

**Review by pcaversaccio — COMMENTED** [review]
`constant` and `immutable` values which are part of the library module are currently not exposed as members when importing the module. The current scope of this PR is to implement `pure`/`view` libraries iiuc. Those values are not part of any state but are part of the runtime bytecode. Thoughts?

**Comment by charles-cooper** [comment]
> `constant` and `immutable` values which are part of the library module are currently not exposed as members when importing the module. The current scope of this PR is to implement `pure`/`view` libraries iiuc. Those values are not part of any state but are part of the runtime bytecode. Thoughts?

i had to cut the scope somewhere and decided to leave those for later since they use the same mechanism as will be used for stateful modules

File: vyper/codegen/stmt.py
Comment (charles-cooper) [reply vyper/codegen/stmt.py:125]: > Moving the calls to a single function could be useful then, like `get_metadata(node, 'key')` or `get_type(node)`, so once it's populated we can change it there

yes, i was just thinking about this! `node.get_type()` could be good as well, it is less verbose. it can also handle that pesky `KeyError` when the type has not been computed yet!

File: vyper/__init__.py
Comment (charles-cooper) [reply vyper/__init__.py]: actually i looked into this and it turns out we already ignore F401 in import files - https://github.com/vyperlang/vyper/blob/0b1f3e143c4f432c469b61fbe1566cb46cfcfca1/setup.cfg#L16. so i was able to just remove the `# noqa: F401` comment

File: tests/conftest.py
Comment (charles-cooper) [reply tests/conftest.py:108]: hmm interesting. well, `InputBundle()` is actually not fully functioning since it's a base class, it will raise exceptions if you try to use it: https://github.com/charles-cooper/vyper/blob/74c630ece2723cf24418acf9c635418ab2c2a461/vyper/compiler/input_bundle.py#L73-L77

so i wanted to use it for tests which need an object of type InputBundle, but can never actually call into it.

**Review by pcaversaccio — COMMENTED** [review]
Another 2 comments after testing:

1. My first comment is about unused imports and devex (it's related to my issue https://github.com/vyperlang/vyper/issues/3272). There is no compilation warning issued currently if I import libraries that are not used. I think it would be a good idea to do that for unused imports in order to improve the quality and cleanliness of the code. This must not be part of that PR, but we should not forget about it.
2. I'm wondering whether we should allow nested module calls:
```vyper
# Lib1.vy

@internal
@view
def blocknumber() -> uint256:
    return block.number + 1

# Lib2.vy

import Lib1

@internal
@view
def blocknumber() -> uint256:
    return Lib1.blocknumber()

# Main.vy

import Lib2

@external
@view
def blocknumber() -> uint256:
    return Lib2.Lib1.blocknumber()
```
Thoughts?

**Review by pcaversaccio — COMMENTED** [review]
Another 2 feedback:

- We should add more tests using Vyper built-in functions.
- The current implementation allows for stateful changes, if you apply some tricks:
```vy
# Lib

@internal
@payable
def call(target: address) -> Bytes[32]:
    response: Bytes[32] = raw_call(self, method_id("hello()"), max_outsize=32, value=msg.value)
    return response

# Main.vy

import Lib

var: uint256


@external
@payable
def hello():
    self.var = msg.value

@external
@payable
def call(target: address) -> Bytes[32]:
    return Lib.call(target)
```

This is more a discussion about the scope of the PR. If it's `pure` and `view` libraries, we should disallow `raw_call`'s IMO.

**Comment by charles-cooper** [comment]
> This is more a discussion about the scope of the PR. If it's `pure` and `view` libraries, we should disallow `raw_call`'s IMO.

no, "stateless" refers more to the fact that importing variables (of any kind) is not supported, not the mutability properties of the imported functions.

**Comment by charles-cooper** [comment]
> My first comment is about unused imports and devex (it's related to my issue https://github.com/vyperlang/vyper/issues/3272). There is no compilation warning issued currently if I import libraries that are not used. I think it would be a good idea to do that for unused imports in order to improve the quality and cleanliness of the code. This must not be part of that PR, but we should not forget about it.

yes this is probably useful, but i'm declaring it out-of-scope for this PR

**Comment by charles-cooper** [comment]
> I'm wondering whether we should allow nested module calls:

fixed in https://github.com/vyperlang/vyper/pull/3663/commits/64bc3a57819dbc4673d9a1394e936e3ebccb4def

**Review by pcaversaccio — APPROVED** [review]
I'm approving with the following caveat: I have tested some of the logic but _not all_, and I also haven't reviewed every single line of code. Under normal circumstances, I wouldn't approve the PR, _BUT_ this PR is an important foundational layer for upcoming PRs that need to be worked on. Furthermore, we have to fully audit/review the stateless and stateful modules implementation, and the sooner we get there (hopefully in January) the better. For anyone going through this PR in the future: Please be aware that there might be bugs, but those bugs will be fixed in upcoming PRs. We don't compromise on any security here, but we have to first get the features done in a proper way, and ensure the complete security based on a feature-complete (in that case library modules) state of the compiler.

File: vyper/semantics/analysis/import_graph.py
Comment (fubuloubu) [reply vyper/semantics/analysis/import_graph.py:27]: Because of the side effect, I think this should be refactored into an if statement that raises an exception of some sort


The bug of missing this statement would be too insidious if the assert was somehow skipped, no matter how unlikely that would be

File: vyper/semantics/types/module.py
Comment (fubuloubu) [reply vyper/semantics/types/module.py:252]: Should the `__eq__` implement something to do with the module?

Comment (fubuloubu) [reply vyper/semantics/types/module.py:252]: This one seems important

**Review by fubuloubu — COMMENTED** [review]
Still some code quality issues to address, but otherwise looks fine 

Happy with @pcaversaccio's manual testing

File: vyper/semantics/types/module.py
Comment (charles-cooper) [reply vyper/semantics/types/module.py:252]: it's outdated -- latest commit has `__eq__` implemented

Comment (charles-cooper) [reply vyper/semantics/types/module.py:252]: it's outdated - latest commit calls `super().__init__()`

**Review by fubuloubu — APPROVED** [review]

**Comment by charles-cooper** [comment]
@pcaversaccio re. builtins see https://github.com/vyperlang/vyper/pull/3663/commits/bea05cafad6f2536eb559028f9cfbf7d7347ac08

**Pull request merged**
